src/Plugin/formatter/FormatterJsonApi.php

```diff
@@ -286,14 +231,80 @@ protected function getRequest() {
   /**
    * Move the embedded resources to the included key.
    *
+   * Change the data structure from an auto-contained hierarchical tree to the
+   * final JSON API structure. The auto-contained tree has redundant information
+   * because every branch contains all the information that is embedded in there
+   * and can be used as stand alone.
+   *
    * @param array $output
    *   The output array to modify to include the compounded documents.
    * @param array $included
    *   Pool of documents to compound.
    *
+   * @return array
+   *   The processed data.
+   *
    * @throws \Drupal\restful\Exception\InternalServerErrorException
    */
-  protected function moveEmbedsToIncluded(array &$output, array $included) {
+  protected function decompress(array $output, array &$included) {
+    static $depth = -1;
+    $depth++;
+    if (!is_array($output)) {
+      // $output is a simple value.
+      $depth--;
+      return $output;
+    }
+    $result = array();
+    if (ResourceFieldBase::isArrayNumeric($output)) {
+      foreach ($output as $item) {
+        $result[] = $this->decompress($item, $included);
+      }
+      $depth--;
+      return $result;
+    }
+    if (!empty($output['__resource__name'])) {
+      $result['type'] = $output['__resource__name'];
+    }
+    if (!empty($output['__resource__id'])) {
+      $result['id'] = $output['__resource__id'];
+    }
+    if (empty($output['__fields'])) {
+      $depth--;
+      return $this->decompress($output, $included);
+    }
+    foreach ($output['__fields'] as $field_name => $field_contents) {
+      if (empty($field_contents['__embedded'])) {
+        $result['attributes'][$field_name] = $field_contents;
+      }
+      else {
+        // Handle single and multiple relationships.
+        $rel = array();
+        $single_item = $field_contents['__cardinality'] == 1;
```

Comment:
Yeah, there is a lot of _proprietary format_ business logic here. I did not know how to avoid this and still have cacheable entities (with all the includes that they may need in the cache record).

This denormalization + optional caching + re-normalization is (IMO) the ugliest part of RESTful. I can see that we can improve this in future sweeps, but I think it's a _good enough_ **first** iteration.